web/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx

```diff
@@ -70,15 +68,14 @@ export default function useTdpClientCanvas(props: Props) {
   const encoder = useRef(new TextEncoder());
   const latestClipboardDigest = useRef('');
 
-  /**
-   * Tracks whether the next keydown or keyup event should sync the
-   * local toggle key state to the remote machine.
-   *
-   * Set to true:
-   * - On component initialization, so keys are synced before the first keydown/keyup event.
-   * - On focusout, so keys are synced when the user returns to the window.
-   */
-  const syncBeforeNextKey = useRef(true);
+  const keyboardHandler = useRef(new KeyboardHandler());
+  useEffect(() => {
+    // On unmount, clear all the timeouts on the keyboardHandler.
+    return () => {
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      keyboardHandler.current.onUnmount();
+    };
+  }, []);
```

Comment:
There's this fragment in React docs:

> Try to [write every Effect as an independent process](https://react.dev/learn/lifecycle-of-reactive-effects#each-effect-represents-a-separate-synchronization-process) and [think about a single setup/cleanup cycle at a time.](https://react.dev/learn/lifecycle-of-reactive-effects#thinking-from-the-effects-perspective) It shouldn’t matter whether your component is mounting, updating, or unmounting. When your cleanup logic correctly “mirrors” the setup logic, your Effect is resilient to running setup and cleanup as often as needed.

Making sure that our code adheres to React's rules will be even more important once something like [the React compiler](https://react.dev/blog/2024/02/15/react-labs-what-we-have-been-working-on-february-2024) drops.

In Strict mode, this effect would be called twice. But the effect is not symmetrical, so this component would break.

Fortunately, I think the fix here is quite straightforward. `keyboardHandler.current` is used only from within key handlers, which pretty much guarantees that none of our code will call the handler until the component actually gets rendered (and thus the effect gets executed).

In this case, we could initialize a new instance of `KeyboardHandler` only within the effect itself and assign it to `keyboardHandler.current`. This way the effect becomes symmetrical and adheres to React's rules.

Reply:
In this case I don't think calling `onUnmout` in the case of a double render would break the component. Still, granted it makes sense to keep these "symmetrical".

63f1dfd6bc7e9a429639225572a0060a0925f1f0